src/goto-instrument/contracts/dynamic-frames/dfcc_infer_loop_assigns.h

```diff
@@ -0,0 +1,34 @@
+/*******************************************************************\
+
+Module: Dynamic frame condition checking
+
+Author: Qinheping Hu, qinhh@amazon.com
+
+Author: Remi Delmas, delmasrd@amazon.com
+
+Date: March 2023
+
+\*******************************************************************/
+
+/// \file
+/// Infer a set of assigns clause targets for a natural loop.
+
+#ifndef CPROVER_GOTO_INSTRUMENT_CONTRACTS_DYNAMIC_FRAMES_DFCC_INFER_LOOP_ASSIGNS_H
+#define CPROVER_GOTO_INSTRUMENT_CONTRACTS_DYNAMIC_FRAMES_DFCC_INFER_LOOP_ASSIGNS_H
+
+#include <analyses/local_may_alias.h>
+#include <goto-instrument/loop_utils.h>
+class source_locationt;
+class messaget;
+class namespacet;
+
+// \brief Infer assigns clause targets for a loop from its instructions and a
+// may alias analysis at the function level.
+std::set<exprt> dfcc_infer_loop_assigns(
```

Comment:
Is a `set` the most appropriate data structure here? Do you actually need the ordering? I'd otherwise suggest to use `std::unordered_set`.

Reply:
this is imposed by the `typedef std::set<exprt> assignst;` typedef defined in `loop_utils.h` and the fact that we return an `assignst`. Returning an unordered set would add an extra `set` to `unordered_set` translation step. I'd rather change the return type to `assignst`.

Reply:
Shouldn't this now be `assignst` as you suggest?

Reply:
fixed.